File: UniGitEsprit/src/main/java/com/esprit/microservice/unigitesprit/services/interfaces/GroupService.java
package com.esprit.microservice.unigitesprit.services.interfaces;

import com.esprit.microservice.unigitesprit.dto.GroupCreateDTO;
import com.esprit.microservice.unigitesprit.dto.GroupResponseDTO;
import com.esprit.microservice.unigitesprit.dto.UserRoleResponseDTO;
import jakarta.mail.MessagingException;

import java.util.List;

public interface GroupService {
    GroupResponseDTO addGroup(GroupCreateDTO groupCreateDTO) ;
    List<GroupResponseDTO> getAllGroups();

    List<GroupResponseDTO> getGroupsByUser(String identifiant);

    GroupResponseDTO getGroupById(Long id);
    GroupResponseDTO updateGroup(Long id, GroupCreateDTO groupCreateDTO);
    void deleteGroup(Long id);
    GroupResponseDTO toggleFavorite(Long id);
     GroupResponseDTO getGroupByPipelineId(Long pipelineId) ;
    GroupResponseDTO addMemberToGroup(Long groupId, UserRoleResponseDTO request);
    GroupResponseDTO removeMemberFromGroup(Long groupId, Long userId);
    GroupResponseDTO updateMemberRole(Long groupId, Long userId, String role);

    List<GroupResponseDTO> searchGroups(String query);
}

File: UniGitEspritFrontV2/src/app/services/group.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { GroupResponse, GroupCreate } from '../models/group.model';

@Injectable({
  providedIn: 'root'
})
export class GroupService {
  private apiUrl = 'http://localhost:8081/uniGitEsprit/api/groups';

  constructor(private http: HttpClient) {}

  addGroup(group: GroupCreate): Observable<GroupResponse> {
    return this.http.post<GroupResponse>(this.apiUrl, group);
  }

  getGroupsByUser(id: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/GroupsbyUser/${id}`);
  }

  getAllGroups(): Observable<GroupResponse[]> {
    return this.http.get<GroupResponse[]>(this.apiUrl);
  }

  getGroupById(id: number): Observable<GroupResponse> {
    return this.http.get<GroupResponse>(`${this.apiUrl}/${id}`);
  }

  updateGroup(id: number, group: GroupCreate): Observable<GroupResponse> {
    return this.http.put<GroupResponse>(`${this.apiUrl}/${id}`, group);
  }

  deleteGroup(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  toggleFavorite(id: number): Observable<GroupResponse> {
    return this.http.post<GroupResponse>(`${this.apiUrl}/${id}/toggle-favorite`, {});
  }
  getGroupByPipelineId(pipelineId: number): Observable<GroupResponse> {
    return this.http.get<GroupResponse>(`${this.apiUrl}/by-pipeline/${pipelineId}`);
  }
  addMember(groupId: number, payload: { userId: number, role: string[] }) {
    return this.http.post<GroupResponse>(`${this.apiUrl}/${groupId}/members`, payload);
  }
  
  removeMember(groupId: number, userId: number) {
    return this.http.delete<GroupResponse>(`${this.apiUrl}/${groupId}/members/${userId}`);
  }
  
  updateMemberRole(groupId: number, userId: number, role: string[]) {
    return this.http.put<GroupResponse>(`${this.apiUrl}/${groupId}/members/${userId}`, role);
  }
  searchGroups(query: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/search?query=${query}`);
  }
}